apisix/plugins/hmac-auth.lua

```diff
@@ -0,0 +1,286 @@
+    local canonical_query_string = ""
+    local request_method = ngx_req.get_method()
+    local args = ngx_req.get_uri_args()
+
+    if type(args) == "table" then
+        local keys = {}
+        local query_tab = {}
+        local query_tab_size = 1
+
+        for k,v in pairs(args) do
+            core.table.insert(keys, k)
+        end
+        core.table.sort(keys)
+
+        for _, key in pairs(keys) do
+            local param = args[key]
+            if type(param) == "table" then
+                for _, vval in pairs(param) do
+                    query_tab[query_tab_size] = escape_uri(key) .. "=" .. escape_uri(vval)
+                    query_tab_size = query_tab_size + 1
+                end
+            else
+                query_tab[query_tab_size] = escape_uri(key) .. "=" .. escape_uri(param)
+                query_tab_size = query_tab_size + 1
+            end
+        end
+        canonical_query_string = core.table.concat(query_tab, "&")
+    end
+
+    local req_body, _ = core.request.get_body()
+    req_body = req_body or ""
+
+    local signing_string = request_method .. canonical_uri ..  canonical_query_string ..
+    req_body .. params.access_key .. params.timestamp .. secret_key
+
+    return hmac_funcs[params.algorithm](secret_key, signing_string)
+end
+
+
+local function validate(ctx, params)
+    if not params.access_key or not params.signature then
+        return false, nil, {message = "access key or signature missing"}
+    end
+
+    local consumer, err = get_consumer(params.access_key)
+    if err then
+        return false, nil, err
+    end
+
+    local conf = consumer.auth_conf
+    if conf.algorithm ~= params.algorithm then
+        return false, nil, {message = str_fmt("algorithm %s not supported", params.algorithm)}
+    end
+
+    core.log.info("conf.clock_skew:", conf.clock_skew)
+    if conf.clock_skew and conf.clock_skew > 0 then
+        local diff = abs(ngx_time() - params.timestamp)
+        core.log.info("conf.diff:", diff)
+        if diff > conf.clock_skew then
+          return false, nil, {message = "Invalid timestamp"}
+        end
+    end
+
+    local secret_key          = conf and conf.secret_key
+    local request_signature   = ngx_decode_base64(params.signature)
+    local generated_signature = generate_signature(ctx, secret_key, params)
+
+    core.log.info("request_signature:", request_signature)
+    core.log.info("generated_signature:", generated_signature)
+
+    return request_signature == generated_signature, consumer
+end
+
+local function get_params(ctx)
+    local params = {}
+    local local_conf = core.config.local_conf()
+    local signature_key = SIGNATURE_KEY
+    local algorithm_key = ALGORITHM_KEY
+    local timestamp_key = TIMESTAMP_KEY
+    local access_key = ACCESS_KEY
+
+    if try_attr(local_conf, "plugin_attr", "hmac-auth") then
+        local attr = local_conf.plugin_attr["hmac-auth"]
+        signature_key = attr.signature_key or signature_key
```

Comment:
this should be a bug. we should not update the variable `signature_key`, it can be used in multiple requests.